FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/AzureTableMetadataStore.cs

```diff
@@ -0,0 +1,169 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Data.Tables;
+using EnsureThat;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Fhir.Synapse.Common.Configurations;
+using Microsoft.Health.Fhir.Synapse.Core.Jobs.Models.AzureStorage;
+using Microsoft.Health.Fhir.Synapse.JobManagement;
+
+namespace Microsoft.Health.Fhir.Synapse.Core.Jobs
+{
+    public class AzureTableMetadataStore : IMetadataStore
+    {
+        private readonly TableClient _metadataTableClient;
+        private readonly ILogger<AzureTableMetadataStore> _logger;
+
+        private const int MaxCountOfQueryEntities = 50;
+
+        // The entities count of a transaction is limited to 100
+        // https://docs.microsoft.com/en-us/rest/api/storageservices/table-service-error-codes
+        private const int MaxCountOfTransactionEntities = 100;
+
+        public AzureTableMetadataStore(
+            IAzureTableClientFactory azureTableClientFactory,
+            IOptions<JobConfiguration> config,
+            ILogger<AzureTableMetadataStore> logger)
+        {
+            EnsureArg.IsNotNull(azureTableClientFactory, nameof(azureTableClientFactory));
+            EnsureArg.IsNotNull(config, nameof(config));
+            EnsureArg.IsNotNullOrWhiteSpace(config.Value.AgentName, nameof(config.Value.AgentName));
+
+            _metadataTableClient = azureTableClientFactory.Create(TableKeyProvider.MetadataTableName(config.Value.AgentName));
+            _metadataTableClient.CreateIfNotExists();
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+        }
+
+        public async Task<TriggerLeaseEntity> GetTriggerLeaseEntityAsync(byte queueType, CancellationToken cancellationToken = default)
+        {
+            return (await _metadataTableClient.GetEntityAsync<TriggerLeaseEntity>(
+                TableKeyProvider.LeasePartitionKey(queueType),
+                TableKeyProvider.LeaseRowKey(queueType),
+                cancellationToken: cancellationToken)).Value;
+        }
+
+        public async Task<Response> AddEntityAsync(ITableEntity tableEntity, CancellationToken cancellationToken = default)
+        {
+            return await _metadataTableClient.AddEntityAsync(tableEntity, cancellationToken);
+        }
+
+        public async Task<Response> UpdateEntityAsync(ITableEntity tableEntity, CancellationToken cancellationToken = default)
+        {
+            return await _metadataTableClient.UpdateEntityAsync(
+                tableEntity,
+                tableEntity.ETag,
+                cancellationToken: cancellationToken);
+        }
+
+        public async Task<CurrentTriggerEntity> GetCurrentTriggerEntityAsync(byte queueType, CancellationToken cancellationToken = default)
+        {
+            CurrentTriggerEntity entity = null;
+            try
+            {
+                var response = await _metadataTableClient.GetEntityAsync<CurrentTriggerEntity>(
```

Comment:
Do we need to further check the response status code here? to make sure the response is successful.